### ISSUE TITLE
Gradient doesn't work!

### ISSUE THREAD

**Issue by nchistov** [fyne-io/fyne#4608]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

I write simple code to try how gradient work, but when I run it, there in the window isn't a gradient from white to black, but only grey color.

### How to reproduce

See the example code.

### Screenshots

![Снимок экрана от 2024-02-04 13-10-29](https://github.com/fyne-io/fyne/assets/96254748/9c52aadc-d9d9-4f00-970d-311f4415956d)

### Example code

```go
package main

import (
	"image/color"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/canvas"
)

func main() {
	myApp := app.New()
	w := myApp.NewWindow("Gradient")

	gradient := canvas.NewHorizontalGradient(color.White, color.Black)
	w.SetContent(gradient)

	w.Resize(fyne.NewSize(100, 100))
	w.ShowAndRun()
}
```

### Fyne version

v2.4.2

### Go compiler version

1.21.5

### Operating system and version

Linux Mint 21.3 Virginia

**Comment by andydotxyz** [comment]
I think this may be because color.White and color.Black are not in the RGBA colourspace. We can probably do something to help with this.

However you could help us test by using `&color.NRGBA{0xff, 0xff, 0xff, 0xff}` and `&color.NRGBA{0x00, 0x00, 0x00, 0x00}` to see if it works.

**Comment by nchistov** [comment]
@andydotxyz No, this isn't working. Window is white.

**Comment by andydotxyz** [comment]
I tried the code and it works here.
I wonder if this can be related to the desktop or window manager somehow.
What happens if you resize the window?

**Comment by andydotxyz** [comment]
I cannot understand how.
Please try v2.4.3 and the develop versions of Fyne to see as somehow specific to 2.4.2.

**Comment by nchistov** [comment]
@andydotxyz on 2.4.3 and 2.4.2 nothing changed.

**Comment by andydotxyz** [comment]
> @andydotxyz on 2.4.3 and 2.4.2 nothing changed.

what about @develop?

**Comment by nchistov** [comment]
@andydotxyz on v2.3.5 also nothing changed.

**Comment by nchistov** [comment]
@andydotxyz on @develop also this (changed only outline color, from white to black):

![Снимок экрана от 2024-02-07 13-58-34](https://github.com/fyne-io/fyne/assets/96254748/e718bb2c-3cc4-45b4-9cb0-2632748d9080)

**Comment by andydotxyz** [comment]
Can anyone on Linux Mint please run the code and verify?
I cannot replicate this on any computer.

**Comment by andydotxyz** [comment]
I think really we will have to dig into what GPU you have, and what graphics driver @nchistov

**Comment by nchistov** [comment]
@andydotxyz my GPU is "AMD Radeon RX 550 Series". And how to see what graphical driver I have?

**Comment by nchistov** [comment]
But I install these deps: `sudo apt-get install golang gcc libgl1-mesa-dev xorg-dev` as in documentation for Debian/Ubuntu linuxes.

**Comment by adamantike** [comment]
I found that removing the caching logic from the [`gl.Painter.getTexture` method](https://github.com/fyne-io/fyne/blob/4f9d61ba12098967e512ececa047cba0365273cd/internal/painter/gl/texture.go#L32) fixed the linear gradient. After debugging, I found that:

1. On gradient creation, the gradient size is 1x1. A single pixel is generated when computing the gradient, and the texture is saved to the cache.
2. Then the window is resized to 100x100, and the gradient changes to 92x92.
3. A redraw is triggered. However, as the texture cache uses the Canvas object as the key, a cache hit is found for the 1x1 texture generated earlier.

I think having canvas objects being able to customize their texture cache key would be ideal. Both for linear and radial gradients, the cache key should also include their current sizes, to avoid cache hits when resized.
